### ISSUE TITLE
Introduce a move command for moving cursor in the editor

### ISSUE THREAD

**Issue by sandy081** [Microsoft/vscode#9143]
Introduce a move command for moving cursor at a logical view position in the editor

This will allow the extension writers to place the cursor at a provided position in the editor

Usage:
- Command Id: `cursorMove`
- Arguments: `{to: $ViewPosition, select: boolean, by: $By, amount: number}`
- Defaults: `amount` =  1. `by` based on `to` value.

**ViewPosition**

- `left`: Moves the cursor left by `n` characters.
- `right`: Moves the cursor right by `n` characters. 
- `up`: Moves the cursor up by `n` lines or wrapped lines.
- `down`: Moves the cursor down by `n` lines or wrapped lines. 

- `wrappedLineStart`: Places the cursor at the start of the current line in the view.
- `wrappedLineFirstNonWhitespaceCharacter`: Places the cursor at the first non white space character of the current line in the view.
- `wrappedLineEnd`: Places the cursor at the end of the current line in the view.
- `wrappedLineLastNonWhitespaceCharacter`: Places the cursor at the last non white space character of the current line in the view.
- `wrappedLineColumnCenter`: Places the cursor at the center of the line in the view.

- `viewPortTop`: Move the cursor to the first non white space character of Nth line from the top of the view.
- `viewPortCenter`: Move the cursor to the first non white space character of the center line of the view.
- `viewPortBottom`: Move the cursor to the first non white space character of Nth line from the bottom of the view.

**By**
- `line` - Default for line specific commands
- `wrappedLine`
- `character` - Default for character specific commands
- `halfLine` - Only for character specific commands.

Set `select` arg to `true`, to select text for above

**Comment by sandy081** [comment]
@rebornix I am going to implement a command name 'move' which can take arguments. Based on the arguments command will move the cursor appropriately. Let me know if you have any specifications so that I can make them part of it.

As a beginning, this will let to move the cursor to
- first character in the screen line
- first non white space character in the screen line

**Comment by sandy081** [comment]
```javascript
// set anchor=active for just positioning the cursor (no selection)
editor.selections = [new vscode.Selection(anchor, active)]
```

@siegebell No, above will place the cursor as per the editor model. Instead this new command will allow you to move it to a position you wanted on the screen. For eg: If a line is wrapped and if you want to move the cursor to the beginning of the line on the screen (not to the beginning of the complete line) above javascript will not able to do that.

**Comment by sandy081** [comment]
@siegebell We can but this will introduce view properties on the document model which is not a good idea. Hence we intended to provide commands so that users can execute them.

**Comment by rebornix** [comment]
@sandy081 these commands look good to me and really thank you! For now I still need time to check whether they are suitable to Vim integration, whether it's a direct support or a workaround.

**Comment by rebornix** [comment]
@sandy081 just played around with these APIs, thank you very much as they make the implementation really simple our side. I bet other extensions can also benefit from these APIs.

There is some unexpected behaviors while running  `cursorMove`, the first thing is `lineEnd` actually moves the cursor the first character of next screen line. The position line and character are both zero-based, so I'm guesssing maybe we forget to decrease by 1 while calculating.

Another one is `lineCenter` is just keeping the cursor where it was. Not sure did I miss anything.

**Comment by sandy081** [comment]
@rebornix I see `lineEnd` and `lineCenter` commands are behaving as expected for me. I guess you would be calling these arguments just similar to others. If you can provide any information that can help in reproducing would be great.

**Comment by rebornix** [comment]
@sandy081 I used the keywords put in this issue description (`up`, `down`), while it's `lineUp` and `lineDown` per your code. It was my misunderstand and `lineCenter`, `lineUp` and `lineDown` works perfectly.

But I still have problem with `lineEnd`, will sync with @sandy081 offline. All others are really awesome.

**Comment by sandy081** [comment]
@rebornix Yah, I forgot to update the description with updated keywords. Thanks for pointing this.

**Comment by sandy081** [comment]
@rebornix It would be great if you can give me an updated list of VIM commands (all variations) which are implemented / implementable (with existing API) and which are still not.  I have seen the VIM extension road map but I think it is not up to date or does not have all variations.. we can also discuss it offline. 

Thanks

**Comment by sandy081** [comment]
@rebornix Commands `H`, `M` and `L` are supported with selection.

**Comment by rebornix** [comment]
@sandy081 I just found that Vim's `H`, `M`, `L` are not referring to Screen Line. They are referring to real line numbers on the gutter. I'm not sure whether it's confusing putting both `screen line` and `real line` in the same API all together, what do you think?

**Comment by sandy081** [comment]
@rebornix above commands are moving to Screen line not to real line. Can you please check by having a very small wrapping size. I think the issue you are seeing is if the screen line above top screen line is little visible then H command jumps to that. Also there is a similar issue for bottom of the view - #9542. It looks to me an issue while computing the lines layout. Since Alex is on vacation we have to wait on that.

Otherwise, `H`, `M` and `L` commands are moving to screen line. Please double check with a line wrapped into more than 2 lines.

**Comment by sandy081** [comment]
@rebornix 

- Re-Implemented `viewPortTop`, `viewPortBottom` and `viewPortCenter` commands by real lines instead of wrapped lines.
- Implement `left` and `right` commands to move cursor left and right respectively by `1`, `n` or `halfLine` characters
- Changed the key word names

Adopt to the above changes and let me know the feedback.

Thanks

**Comment by sandy081** [comment]
@rebornix Please note the small change done to the API for consistency and clean

- Renamed argument parameter `amount` to `value`.

Please adopt to these chages

**Comment by aminroosta** [comment]
![vkkkkkk](https://cloud.githubusercontent.com/assets/2920178/18233186/3e825d32-72f5-11e6-8970-fbfbe9bc84bc.gif)

@sandy081 `cursorMove` does not respect already selected text.
`Vjjjj` ing works find but `Vkkkkkk` ing will break selected code.

```js
    await vscode.commands.executeCommand("cursorMove", {
     to: "up",
     by: "wrappedLine",
     select: true,
     value: count
   });
```
I've spent a couple of hours on this, it seems that `activeTextEditor.slection` is not handled properly.

@sandy081 Please give us the same API without moving the cursor, Vim extensions are masters on how to move the cursor themselves.

**Comment by aminroosta** [comment]
@sandy081  This is how we can practically use your api, `moveDown` record the position then `moveUp`. This makes life hard and moving around slow and if you move pass the first line or the last line some wired visual feedback can be seen (you can see cursor moving up and then down and them up) ❗ 

Let's face it, no other plugin is going to use these commands other than VIM extensions 😉 
And we already know that no internal editor stuff needs to be exposed for folding info.

These makes both vim extension writer life and vscode team easier.
So i'm going to ask once more hoping that you reconsider this wrong design approach 😄

**Comment by johnfn** [comment]
@sandy081 

I would like to chime in along with @aminroosta here.

Unfortunately, the current cursorMove API is not powerful enough to implement most core features of Vim. Simply moving the cursor is not enough - we actually need the position. Without the position, there's no practical way to use the value any of the visual modes (which effectively take the position, then apply some transformations on it, and then create a new selection object). This means that visual mode, visual line mode and visual block mode cannot be implemented using this API.

Again, we are all super appreciative of all the work that you guys do on such a great editor. Let us know if there's anything we can do to help!

**Comment by sandy081** [comment]
@aminroosta 

>> @sandy081 `cursorMove` does not respect already selected text.
`Vjjjj` ing works find but `Vkkkkkk` ing will break selected code.

It is working for me very smoothly. Please see the attached video. I call the following command just like you did

```js
vscode.commands.executeCommand("cursorMove", {
     to: "up",
     by: "wrappedLine",
     select: true,
     value: count
   });
```

![sep-05-2016 18-15-05](https://cloud.githubusercontent.com/assets/10746682/18253437/edabc7c2-7394-11e6-8a56-e4de5e9a1fa5.gif)

>> Unfortunately, the current cursorMove API is not powerful enough to implement most core features of Vim. 

@rebornix, In regards to @johnfn comment,  may I know what is missing here? I remember that this can be used to implement most of cursor movement functionality in Vim. In fact, I implemented in our [Vim sample extension] (https://github.com/Microsoft/vscode-extension-samples/tree/master/vim-sample).

**Comment by aminroosta** [comment]
@sandy081 Thanks for taking time and checking this issue. So there should be something wrong with my code. If your code is available online on github please let me know.

**Comment by sandy081** [comment]
@aminroosta here is our [Vim sample extension] (https://github.com/Microsoft/vscode-extension-samples/tree/master/vim-sample).

**Comment by aminroosta** [comment]
@sandy081 No no no, `vkkkk` (lower case `v`) is perfectly fine, i'm having a hard time with `Vkkkk` (upper case `V`) that selects lines.

**Comment by aminroosta** [comment]
![vkkkk](https://cloud.githubusercontent.com/assets/2920178/18276572/ae70dc38-7460-11e6-8c21-cc9fd681a19a.gif)

@sandy081 It seems only lower case `vk` can be implemented with `cursorMove`, i can't fine any information in the documentation (first comment above) that suggests how to implement `Vkkk` ing.

**Comment by sandy081** [comment]
@aminroosta Just to note that the sample vim extension does not have all complete vim features. This is an example/reference implementation showing how some of the vim features can be implemented using the editor commands. 

With regards to `upper case V`, sample supports moving to visual mode only on `lower case v`. You can refer to this to implement the other.

@rebornix might help you in how to consume this.

**Comment by rebornix** [comment]
@sandeep thanks for the tip.

@aminroosta I'm adding these command sin my separate PR but it does hacking to the view updating phase so it's still WIP. I'll let you know if I solve the issues you met.

**Comment by rebornix** [comment]
I see the problem, I only used `cursorMove` for screen lines moving but not regular lines, that might describe I don't see any issues @aminroosta ran into. But @aminroosta is doing the right thing, implementing all Vim cursor moving commands by `cursorMove` instead of doing things in different ways. Let me give them a check real quick.
